scrapy/core/http2/stream.py

```diff
@@ -0,0 +1,422 @@
+
+        self._cb_close(self.stream_id)
+        self.stream_closed_server = True
+
+        flags = None
+        if b'Content-Length' not in self._response['headers']:
+            # Missing Content-Length - {twisted.web.http.PotentialDataLoss}
+            flags = ['partial']
+
+        # NOTE: Order of handling the events is important here
+        # As we immediately cancel the request when maxsize is exceeded while
+        # receiving DATA_FRAME's when we have received the headers (not
+        # having Content-Length)
+        if reason is StreamCloseReason.MAXSIZE_EXCEEDED:
+            expected_size = int(self._response['headers'].get(b'Content-Length', -1))
+            error_msg = (
+                f'Cancelling download of {self._request.url}: expected response '
+                f'size ({expected_size}) larger than download max size ({self._download_maxsize}).'
+            )
+            logger.error(error_msg)
+            self._deferred_response.errback(CancelledError(error_msg))
+
+        elif reason is StreamCloseReason.ENDED:
+            self._fire_response_deferred(flags)
+
+        # Stream was abruptly ended here
+        elif reason is StreamCloseReason.CANCELLED:
+            # Client has cancelled the request. Remove all the data
+            # received and fire the response deferred with no flags set
+
+            # NOTE: The data is already flushed in Stream.reset_stream() called
+            # immediately when the stream needs to be cancelled
+
+            # There maybe no :status in headers, we make
+            # HTTP Status Code: 499 - Client Closed Request
+            self._response['headers'][':status'] = '499'
+            self._fire_response_deferred()
+
+        elif reason in (StreamCloseReason.RESET, StreamCloseReason.CONNECTION_LOST):
+            self._deferred_response.errback(ResponseFailed([
+                error if error else Failure()
+            ]))
+
+        elif reason is StreamCloseReason.INACTIVE:
+            self._deferred_response.errback(InactiveStreamClosed(self._request))
+
+        elif reason is StreamCloseReason.INVALID_HOSTNAME:
+            self._deferred_response.errback(InvalidHostname(
+                self._request,
+                self._conn_metadata['hostname'],
+                f'{self._conn_metadata["ip_address"]}:{self._conn_metadata["port"]}'
+            ))
+
+    def _fire_response_deferred(self, flags: List[str] = None):
+        """Builds response from the self._response dict
+        and fires the response deferred callback with the
+        generated response instance"""
+
+        body = self._response['body'].getvalue()
+        response_cls = responsetypes.from_args(
+            headers=self._response['headers'],
+            url=self._request.url,
+            body=body
+        )
+
+        response = response_cls(
+            url=self._request.url,
+            status=self._response['headers'][':status'],
```

Comment:
Should this be cast to `int`? IIRC all headers are strings, and above there is a line that does `self._response['headers'][':status'] = '499'`

Reply:
Right. I'll add a typecast to `int` explicitly here. I did not do that initially because the `Response` already typecast's `status` to `int`.

https://github.com/scrapy/scrapy/blob/56a6d22352071208637abd86bd7b6d81a50d5bdb/scrapy/http/response/__init__.py#L23